buster/chatbot.py

```diff
@@ -121,59 +122,70 @@ def rank_documents(
 
         return matched_documents
 
-    def prepare_prompt(self, question: str, candidates: pd.DataFrame) -> str:
-        """
-        Prepare the prompt with prompt engineering.
-        """
-
-        max_chars = self.cfg.max_chars
-        text_before_prompt = self.cfg.text_before_prompt
-
-        documents_list = candidates.text.to_list()
+    @classmethod
+    def prepare_documents(cls, matched_documents: pd.DataFrame, max_words: int) -> str:
+        # gather the documents in one large plaintext variable
+        documents_list = matched_documents.text.to_list()
         documents_str = " ".join(documents_list)
-        if len(documents_str) > max_chars:
+
+        # truncate the documents to fit
+        # TODO: increase to actual token count
+        word_count = len(documents_str.split(" "))
+        if word_count > max_words:
             logger.info("truncating documents to fit...")
-            documents_str = documents_str[0:max_chars]
+            documents_str = " ".join(documents_str.split(" ")[0:max_words])
+            logger.info(f"Documents after truncation: {documents_str}")
 
-        return documents_str + text_before_prompt + question
+        return documents_str
 
-    def generate_response(self, prompt: str, matched_documents: pd.DataFrame) -> str:
+    def prepare_prompt(
+        self,
+        question: str,
+        matched_documents: pd.DataFrame,
+        text_before_prompt: str,
+        text_before_documents: str,
+    ) -> str:
         """
-        Generate a response based on the retrieved documents.
+        Prepare the prompt with prompt engineering.
         """
-        if len(matched_documents) == 0:
-            # No matching documents were retrieved, return
-            response_text = "I did not find any relevant documentation related to your question."
-            return response_text
+        documents_str: str = self.prepare_documents(matched_documents, max_words=self.cfg.max_words)
+        return text_before_documents + documents_str + text_before_prompt + question
 
-        logger.info(f"querying GPT...")
-        logger.info(f"Prompt:  {prompt}")
+    @classmethod
+    def get_gpt_response(cls, **completion_kwargs):
         # Call the API to generate a response
+        logger.info(f"querying GPT...")
         try:
-            completion_kwargs = self.cfg.completion_kwargs
-            completion_kwargs["prompt"] = prompt
-            response = openai.Completion.create(**completion_kwargs)
-
-            # Get the response text
-            response_text = response["choices"][0]["text"]
-            logger.info(f"GPT Response:\n{response_text}")
-            return response_text
+            return openai.Completion.create(**completion_kwargs)
 
         except Exception as e:
             # log the error and return a generic response instead.
             import traceback
 
             logger.error("Error connecting to OpenAI API")
             logging.error(traceback.format_exc())
```

Comment:
1. Do not do an `except Exception`. Catch the ones that are truly temporary failures.
2. `logger.exception` should be better.

Reply:
that's the thing, I'm not sure which exceptions we can catch, openAI's API is a bit sparse and I've been catching different ones as i've been debugging

good call on the `logger.exception`, fixing now